test/source-sink-errors/mzcompose.py

```diff
@@ -0,0 +1,172 @@
+        c.up("testdrive", persistent=True)
+        c.start_and_wait_for_tcp(services=["redpanda", "materialized", "storaged"])
+        c.wait_for_materialized()
+
+        with c.override(
+            Testdrive(
+                no_reset=True,
+                seed=seed,
+                entrypoint_extra=["--initial-backoff=1s", "--backoff-factor=0"],
+            )
+        ):
+            self.populate(c)
+            self.breakage(c, seed)
+            self.assert_error(c, self.expected_error)
+
+            if self.fixage:
+                self.fixage(c, seed)
+                self.assert_recovery(c)
+
+    def populate(self, c: Composition) -> None:
+        # Create a source and a sink
+        c.testdrive(
+            dedent(
+                """
+                > CREATE CONNECTION kafka_conn
+                  TO KAFKA (BROKER '${testdrive.kafka-addr}');
+
+                > CREATE CONNECTION IF NOT EXISTS csr_conn TO CONFLUENT SCHEMA REGISTRY (
+                  URL '${testdrive.schema-registry-url}'
+                  );
+
+                $ kafka-create-topic topic=source-topic
+
+                $ kafka-ingest topic=source-topic format=bytes
+                ABC
+
+                > CREATE SOURCE source1
+                  FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-source-topic-${testdrive.seed}')
+                  FORMAT BYTES
+                  ENVELOPE NONE
+                  /* WITH ( REMOTE 'storaged:2100' ) REMOTE causes the status source to be empty */
+
+                > CREATE SINK sink1 FROM source1
+                  INTO KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-sink-topic-${testdrive.seed}')
+                  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+                  ENVELOPE DEBEZIUM
+                  /* WITH ( REMOTE 'storaged:2100' ) REMOTE causes the status source to be empty */
+
+                # Wait for the source to start up
+                > SELECT COUNT(*) FROM source1
+                1
+                """
+            )
+        )
+
+    def assert_error(self, c: Composition, error: str) -> None:
+        # Validate that cluster2 continues to operate
+        c.testdrive(
+            dedent(
+                f"""
+                > SELECT status, error ~* '{error}'
+                  FROM mz_internal.mz_source_status
+                  WHERE name = 'source1'
+                stalled true
+
+                #> SELECT status, error ~* '{error}'
+                #  FROM mz_internal.mz_sink_status
+                #  WHERE name = 'sink1'
+                # stalled true
+                """
+            )
+        )
+
+    def assert_recovery(self, c: Composition) -> None:
+        c.testdrive(
+            dedent(
+                """
+                $ kafka-ingest topic=source-topic format=bytes
+                ABC
+
+                > SELECT COUNT(*) FROM source1;
+                2
+
+                #> SELECT status, error
```

Comment:
@bkirwi After the disruption has been cleared, the source is back up and ingesting , as evidenced by the `SELECT COUNT(*) = 2` check above. However, the status does not always transition to "running".

Reply:
Investigating this; might be a timing issue.

Reply:
I think I've figured this one out: an event ordering issue in a multi-worker environment. I suspect most of my local testing was on a single worker.

I'll work up a fix to this, though I'm not sure exactly how to approach it yet.